### ISSUE TITLE
Provide a cleaner async validation implementation

### ISSUE THREAD

**Issue by cbovis** [ctavan/express-validator#249]
The current implementation of async validation results in some pretty messy code. 

Throwing an exception to handle validation errors feels wrong since they're not exceptions, they're expectations. More on this here... http://stackoverflow.com/questions/729379/why-not-use-exceptions-as-regular-flow-of-control

An example of current implementation,

```
Place.findById(req.params.placeId).then(place => {
  return req.asyncValidationErrors().then(() => {
    res.render('places/edit', {
      place
    });
  }).catch(err => {
    if (Array.isArray(err)){
      res.render('places/edit', {
        place,
        errors: err
      });
    } else {
      throw err;
    }
  });
}).catch(err => { next(err); });
```

Suggested implementation,

```
Promise.join(
  Place.findById(req.params.placeId),
  req.validate(),
  (place, errors) => {
    res.render('places/edit', {
      place,
      errors
    });
  }
).catch(err => { next(err); });
```

Using yield syntax via co-express the second method would produce even cleaner code as opposed to the messy try-catch solution that is presently required,

```
var place = yield Place.findById(req.params.placeId);
var errors = yield req.validate();

res.render('places/edit', {
  place,
  errors
});
```

By passing the errors back via the promise resolution you can see the code becomes much cleaner. If you're interested in implementing this then I'm happy to put together a pull request for submission.

Thoughts?

**Comment by gustavohenke** [comment]
I don't see your idea only as a way to make the code cleaner, but also more consistent.
When you do sync validation, express-validator will not throw exceptions. So why throw them when going async?

**Comment by gustavohenke** [comment]
Additionally, as per [semver](http://semver.org/), this would require a major version bump.

**Comment by cbovis** [comment]
I have some spare time right now whilst looking for work so I'm going to fork and take a look at implementing this 👍

**Comment by cbovis** [comment]
Implemented via pull request #269

**Comment by gustavohenke** [comment]
Closed by #269 😉
